[SwiftUI] Conform WebPage to Transferable

rr-codes opened the pull request:
#### 5bc842de99ab753aaca529954c45711dbbed7bbe
<pre>
[SwiftUI] Conform WebPage to Transferable
<a href="https://bugs.webkit.org/show_bug.cgi?id=294528">https://bugs.webkit.org/show_bug.cgi?id=294528</a>
<a href="https://rdar.apple.com/152695298">rdar://152695298</a>

Reviewed by Aditya Keerthi.

`WebPage` has several ad-hoc functions to produce web archives, PDFs, and images. Instead of these
functions, this PR conforms `WebPage` to `Transferable`, which offers several conveniences
and enhanced functionality, as well as being a substantially more modern API design.

This change involves three major parts:

1) Conform `WebPage` to the `Transferable` protocol. This provides a lot of conveniences to the developer;
in particular, it will now be possible to directly use a WebPage directly when performing operations that
are done on `Transferable` types such as drag and drop and copy and paste. Additionally, this makes it possible
for new exportable types to be added in WebKit without having to make any API modifications. Developers can also
directly use the conformance by using the `exported(as:)` function, providing uniformity both within WebKit itself
and amongst other types that conform to `Transferable`.

2) The current `WebPage`/`WKWebView` API provides several configuration options for generating PDFs and images.
If `WebPage` were to just conform to `Transferable`, these customization options would be lost. To maintain parity,
an overload of `exported(as:)` is provided by WebKit itself, with the parameter describing the exported type along
with its configuration. Data created using the normal `Transferable` conformance will use the default configurations.

3) Remove the existing functions from `WebPage`, since they will be obviated by the `Transferable` conformance.

As an implementation detail, `WebPage` can also add support for several other content types, such as URL, rich text,
and link presentation metadata, in addition to pdf, web archive, and image. This does not directly impact the API
surface but is good to note.

* Source/WebKit/UIProcess/API/Swift/WebPage+Transferable.swift: Added.

Conform WebPage to Transferable; to do so, WebPage implements the `transferRepresentaion` rquirement. This provides
different content types that a WebPage can be exported to; specifically, web archive data, pdf, image, and URL.

* Source/WebKit/UIProcess/API/Swift/WebPage.swift:

When a WebPage is used without a WebView, it needs to have some non-zero size so that representations such as PDF
and image can actually be created. If a WebPage is not bound to a WebView, it&apos;s frame is now set to a fixed size.

Also, remove the old PDF and web archive data functions.

* Source/WebKit/_WebKit_SwiftUI/API/WebPage+SwiftUI.swift:

Remove the old snapshot function.

* Tools/SwiftBrowser/Source/ViewModel/BrowserViewModel.swift:

Update the test app.

* Tools/TestWebKitAPI/Tests/WebKit Swift/WebPageTransferableTests.swift: Added.

Add several tests to test the Transferable functionality.

Canonical link: <a href="https://commits.webkit.org/296332@main">https://commits.webkit.org/296332@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/336ece80d4d00d963fbd77547a48aff48c9ba564

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108129 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/27792 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/18212 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/113340 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/58623 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/110092 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/28488 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/36342 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/82101 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/111077 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22573 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/97409 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/62533 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/21990 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/15545 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58083 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/91936 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/15603 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/116465 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/35194 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/25915 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/91126 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/35569 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/93686 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/90921 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23187 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/35810 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13572 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/31002 "Build was cancelled. Recent messages:Printed configuration") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35095 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/40650 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/34829 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38187 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/36496 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/API/Swift/WebPage+Transferable.swift:
```diff
@@ -0,0 +1,257 @@
+public import CoreTransferable
+import Foundation
+import UniformTypeIdentifiers
+@_weakLinked internal import SwiftUI
```
Weak linking, but I don't see any use of `#_hasSymbol`. Wouldn't we observe crashes if someone were to use this API in a place where SwiftUI was missing?

rr-codes replied:
that would essentially only happen if the API was used in the base system where SwiftUI is not built in, but I'll add the _hasSymbol

pxlcoder replied:
I think we should, since `WebPage` can be used without `SwiftUI`.

pxlcoder commented on Source/WebKit/UIProcess/API/Swift/WebPage+Transferable.swift:
```diff
@@ -0,0 +1,257 @@
+        /// A configuration of a webpage for a representation as image data.
+        ///
+        /// - Parameters:
+        ///   - region: The region of the page to generate the image for.
```
```suggestion
        ///   - region: The region of the page used to generate the image.
```

(thoughts? ending a sentence in `for` is a little informal)

rr-codes replied:
in_for_mal lol

pxlcoder commented on Source/WebKit/UIProcess/API/Swift/WebPage.swift:
```diff
@@ -79,6 +79,9 @@ final public class WebPage {
         case notInFullscreen
     }
+    
+    // Arbitrarily chosen; this is a common "default" window size.
```
I think we need a better comment for an API behavior than "arbitrarily chosen".

cc @hortont424 who I believe suggested this value?

rr-codes committed:
[SwiftUI] Conform WebPage to Transferable
https://bugs.webkit.org/show_bug.cgi?id=294528
rdar://152695298

Reviewed by Aditya Keerthi.

`WebPage` has several ad-hoc functions to produce web archives, PDFs, and images. Instead of these
functions, this PR conforms `WebPage` to `Transferable`, which offers several conveniences
and enhanced functionality, as well as being a substantially more modern API design.

This change involves three major parts:

1) Conform `WebPage` to the `Transferable` protocol. This provides a lot of conveniences to the developer;
in particular, it will now be possible to directly use a WebPage directly when performing operations that
are done on `Transferable` types such as drag and drop and copy and paste. Additionally, this makes it possible
for new exportable types to be added in WebKit without having to make any API modifications. Developers can also
directly use the conformance by using the `exported(as:)` function, providing uniformity both within WebKit itself
and amongst other types that conform to `Transferable`.

2) The current `WebPage`/`WKWebView` API provides several configuration options for generating PDFs and images.
If `WebPage` were to just conform to `Transferable`, these customization options would be lost. To maintain parity,
an overload of `exported(as:)` is provided by WebKit itself, with the parameter describing the exported type along
with its configuration. Data created using the normal `Transferable` conformance will use the default configurations.

3) Remove the existing functions from `WebPage`, since they will be obviated by the `Transferable` conformance.

As an implementation detail, `WebPage` can also add support for several other content types, such as URL, rich text,
and link presentation metadata, in addition to pdf, web archive, and image. This does not directly impact the API
surface but is good to note.

* Source/WebKit/UIProcess/API/Swift/WebPage+Transferable.swift: Added.

Conform WebPage to Transferable; to do so, WebPage implements the `transferRepresentaion` rquirement. This provides
different content types that a WebPage can be exported to; specifically, web archive data, pdf, image, and URL.

* Source/WebKit/UIProcess/API/Swift/WebPage.swift:

When a WebPage is used without a WebView, it needs to have some non-zero size so that representations such as PDF
and image can actually be created. If a WebPage is not bound to a WebView, it's frame is now set to a fixed size.

Also, remove the old PDF and web archive data functions.

* Source/WebKit/_WebKit_SwiftUI/API/WebPage+SwiftUI.swift:

Remove the old snapshot function.

* Tools/SwiftBrowser/Source/ViewModel/BrowserViewModel.swift:

Update the test app.

* Tools/TestWebKitAPI/Tests/WebKit Swift/WebPageTransferableTests.swift: Added.

Add several tests to test the Transferable functionality.

Canonical link: https://commits.webkit.org/296332@main
